simx86: generate code (with helper call) for protected mode long jmp/call/ret instructions

bartoldeman opened the pull request:
Expands on #2559 to also include CS. Now that generated code can change `TheCPU.mode`, I also had to make sure it's read back into `basemode` after a `CODE_FLUSH()`.

The rest is quite similar to existing real mode generation but we need to be careful setting CS first before doing anything else in case of exceptions. So the helper now returns the old CS value so it can be pushed *after* the new CS was checked and set if ok. And for `RETl` we need to pop into temporary storage twice to obtain the new CS value, before `TheCPU.esp` can be adjusted etc.

The remaining callers of `MAKESEG` also do so for vm86 (INT and IRET). They are a little harder to deal with.

bartoldeman committed:
simx86: interp: allow for changing TheCPU.mode by exec.

Easiest way is to let CODE_FLUSH assign to basemode, and use
TheCPU.mode directly in _Interp86(). CODE_FLUSH2 doesn't need
to assign as the new mode isn't used by any following calls
of CODE_FLUSH2.

Interp_LONG_CS is already assigned to at all the necessary places,
(after exec and in MAKESEG), no need to also do it in _Interp86().

bartoldeman committed:
simx86: use Gen for jmp/call far absolute in protected mode

The trick for the helper is to return the old CS, and -1 in case
an exception was set. Then the old CS can be pushed if needed
after the new one has been checked for validity and set.

Do the same thing in A_SR_SH4 so the sequence in interp.c can be
more similar for rm and pm.

bartoldeman committed:
simx86: use Gen for call/jmp indirect in protected mode

bartoldeman committed:
simx86: use Gen for RETl in protected mode

Needed a little adjustment to O_POP2: Ofs_RZERO means not to pop
into a register; MPOPRM means to store the stack pointer in %%esi
to avoid it getting modified by CPatch.

bartoldeman committed:
simx86: use Gen for RETlisp in protected mode

bartoldeman commented:
tasm32 failed in CI.

stsp commented on src/base/emu-i386/simx86/codegen-x86.c:
```diff
@@ -303,8 +311,8 @@ static unsigned char *CodeGen_x86(unsigned char *CodePtr, unsigned char *BaseGen
-		// jz skip
-		G2M(0x74,TAILSIZE,Cp);
+		// jns skip
+		G2M(0x79,TAILSIZE,Cp);
```
What is this?
A bugfix?

bartoldeman replied:
It is related to the `SetSegProt` helper returning `-1` for an invalid segment and a positive value for the old segment value.

Before it was `0` for valid and `TheCPU.err2` for "invalid", but the generated code doesn't need to set `TheCPU.err2`.

stsp commented on src/base/emu-i386/simx86/interp.c:
```diff
@@ -347,14 +362,23 @@ static unsigned int _JumpGen(unsigned int P2, int mode, int opc,
 		unsigned short jcs = FetchW(P2 + pskip - 2);
-		Gen(L_REG, mode|DATA16, Ofs_CS);
-		if (!(mode & DATA16)) // 32-bit segreg padding
-		    Gen(L_ZXAX, mode);
```
So no padding any more?

bartoldeman replied:
padding of the old CS is now done in codegen-x86.c:265 under A_SR_SH4.
Eventually I'd like to get rid of L_ZXAX and always read into %eax in the JIT generated code using MOVZX instead of mov ..,%ax, which also avoids partial register stall issues on many CPUs.

bartoldeman committed:
simx86: fix invalid reads in _JumpGen() for indirect jumps

Some indirect jumps were reading the bytes after P0 and not using
them. This was always wrong but only showed up with tasm32.exe
doing a RETF at the end of a page in protected mode, which then
faulted.

bartoldeman commented:
It turned out the tasm32.exe issue in the CI was an older issue: `JumpGen()` for `RETl` was accidentally reading but not using the bytes after the instruction as if it were an direct jump with an immediate value. That is usually harmless except if the retf instruction is at the end of a page and bytes beyond that can't be read.

stsp commented on src/base/emu-i386/simx86/interp.c:
```diff
@@ -1972,17 +1972,22 @@ intop3b:		{ int op = ArOpsFR[D_MO(opc)];
+			    /* pop from stack without adjusting esp */
+			    Gen(O_POP1, _mode);
+			    Gen(O_POP2, _mode, Ofs_RZERO);
+			    Gen(O_POP2, _mode, Ofs_RZERO);
```
What does this do, why
such a difference for PM?

bartoldeman replied:
essentially this does (in 32 bit mode):
```mov (%esp+4), %eax```
the same as the `NOS_WORD` inline functions (so alternatively I could generate a MODRM memory address like this, and load from there, but that's complicated with stackmask etc).
Then this `%eax` is checked if it's valid, *before* anything in `TheCPU` is modified, as this instruction is restartable.

In real address mode the segment is always valid so you can simply pop into ip before knowing what the cs value is.

stsp replied:
I wonder if you can then
avoid the subsequent POPs
and use the tmp locations
rather than RZERO here.
As otherwise this really looks
strange - too many POPs.
But I suppose this was just
the simplest path, if not the
most optimal one.

bartoldeman replied:
`Ofs_RZERO` is just a placeholder to note "tmp".
It may be cleaner to introduce a new mode flag here like MNOREG, and then use
```
Gen(O_POP2, _mode|MNOREG);
```
? Then codegen-x86 and sim would need to check for that instead of Ofs_RZERO.

bartoldeman replied:
but yes the old way (simulator) wasn't as POP-heavy, it had this
```
NOS_WORD
POP
POP_ONLY
```
where
```
static __inline__ void NOS_WORD(int m, uint16_t *w)             // for segments
{
        unsigned int sp = (TheCPU.esp+(m&DATA16? 2:4)) & TheCPU.StackMask;
        GetSWord(w);
}

static __inline__ void POP_ONLY(int m)
{
        unsigned int sp = TheCPU.esp + (m&DATA16? 2:4);
        TheCPU.esp = (sp&TheCPU.StackMask) | (TheCPU.esp&~TheCPU.StackMask);
}
```
It's just a way of finding a good way to do "NOS_WORD".

bartoldeman replied:
Actually I think I see what you mean. It may be possible to use `TheCPU.eip` to store the new eip since even if a GPF occurs, it'll be computed after the fact in interp.c:
```
TheCPU.eip = ret - Interp_LONG_CS;
```
then one could use something like:
```
                            Gen(O_POP1, _mode);
                            Gen(O_POP2, _mode, Ofs_EIP);
                            Gen(O_POP2, _mode, Ofs_RZERO);
                            AddrGen(A_SR_PROT, _mode, Ofs_CS, P0); // or A_SR_SH4 for real addresses
                            Gen(O_POP3, _mode);
```
as the `O_POP3` writes ESP, which would not happen with an exception.

stsp commented:
Thanks!
